File: release-artifacts/v4-external-evidence-handoff/fixtures/external-engine-baselines/v4/unreal/v4_external_visual_baseline_runner.py
"""
Generic Unreal Python scaffold for Galileo3D V4 external visual baseline slots.
Run inside a real Unreal Editor Python session. This builds deterministic proxy
geometry for the descriptor's baselineKind; the resulting screenshot/report are
the evidence, not this scaffold.
"""
import json
import math
import os
import sys
import time
import unreal

SCENE_PATH = sys.argv[1] if len(sys.argv) > 1 else "fixtures/external-engine-baselines/v4/product-visual-parity-scene.json"
SCREENSHOT_PATH = sys.argv[2] if len(sys.argv) > 2 else "tests/reports/v4-external-baseline/unreal-baseline.png"
MESH_PATHS = {
    "cube": "/Engine/BasicShapes/Cube.Cube",
    "box": "/Engine/BasicShapes/Cube.Cube",
    "sphere": "/Engine/BasicShapes/Sphere.Sphere",
    "cylinder": "/Engine/BasicShapes/Cylinder.Cylinder",
}
MATERIAL_CACHE = {}

with open(SCENE_PATH, "r", encoding="utf-8") as scene_file:
    scene = json.load(scene_file)

baseline_kind = scene.get("baselineKind", "product-visual")
unreal.EditorLevelLibrary.new_level("/Game/Galileo3D_V4_" + baseline_kind.replace("-", "_"))

def sanitize_asset_name(value):
    return "".join(ch if ch.isalnum() or ch == "_" else "_" for ch in value)[:48] or "material"

def load_mesh(geometry):
    mesh = unreal.EditorAssetLibrary.load_asset(MESH_PATHS.get(geometry, MESH_PATHS["cube"]))
    if mesh is None:
        raise RuntimeError("Unable to load Unreal built-in mesh for geometry: " + str(geometry))
    return mesh

def material_descriptor(material_id):
    for material in scene.get("materials", []):
        if material.get("id") == material_id:
            return material
    return {"id": material_id or "default", "color": [0.78, 0.78, 0.78, 1.0], "metallic": 0.0, "roughness": 0.5}

def material_for_descriptor(descriptor):
    material_id = descriptor.get("id", "default")
    if material_id in MATERIAL_CACHE:
        return MATERIAL_CACHE[material_id]
    fallback = unreal.EditorAssetLibrary.load_asset("/Engine/BasicShapes/BasicShapeMaterial.BasicShapeMaterial")
    color = descriptor.get("color", [0.78, 0.78, 0.78, 1.0])
    metallic = float(descriptor.get("metallic", 0.0))
    roughness = float(descriptor.get("roughness", 0.5))
    try:
        package_path = "/Game/Galileo3D_V4_ExternalBaseline/Materials"
        unreal.EditorAssetLibrary.make_directory(package_path)
        asset_name = "M_" + sanitize_asset_name(material_id)
        existing = unreal.EditorAssetLibrary.load_asset(package_path + "/" + asset_name + "." + asset_name)
        material = existing or unreal.AssetToolsHelpers.get_asset_tools().create_asset(
            asset_name,
            package_path,
            unreal.Material,
            unreal.MaterialFactoryNew(),
        )
        base = unreal.MaterialEditingLibrary.create_material_expression(material, unreal.MaterialExpressionConstant4Vector, -420, -120)
        base.set_editor_property("constant", unreal.LinearColor(float(color[0]), float(color[1]), float(color[2]), float(color[3]) if len(color) > 3 else 1.0))
        metal = unreal.MaterialEditingLibrary.create_material_expression(material, unreal.MaterialExpressionConstant, -420, 40)
        metal.set_editor_property("r", metallic)
        rough = unreal.MaterialEditingLibrary.create_material_expression(material, unreal.MaterialExpressionConstant, -420, 160)
        rough.set_editor_property("r", roughness)
        unreal.MaterialEditingLibrary.connect_material_property(base, "", unreal.MaterialProperty.MP_BASE_COLOR)
        unreal.MaterialEditingLibrary.connect_material_property(metal, "", unreal.MaterialProperty.MP_METALLIC)
        unreal.MaterialEditingLibrary.connect_material_property(rough, "", unreal.MaterialProperty.MP_ROUGHNESS)
        unreal.MaterialEditingLibrary.recompile_material(material)
    except Exception as error:
        unreal.log_warning("Galileo3D V4 baseline material fallback for " + material_id + ": " + str(error))
        material = fallback
    MATERIAL_CACHE[material_id] = material
    return material

def fallback_material(material_id, hue_index):
    colors = [
        [0.1, 0.14, 0.18, 1.0],
        [0.9, 0.48, 0.18, 1.0],
        [0.08, 0.32, 0.52, 1.0],
        [0.82, 0.78, 0.68, 1.0],
        [0.22, 0.48, 0.9, 1.0],
        [0.9, 0.24, 0.38, 1.0],
        [0.18, 0.78, 0.48, 1.0],
    ]
    return material_for_descriptor({
        "id": material_id,
        "color": colors[hue_index % len(colors)],
        "metallic": 0.08 + (hue_index % 4) * 0.12,
        "roughness": 0.28 + (hue_index % 5) * 0.09,
    })

def spawn_actor(name, location, scale, geometry="cube", material=None):
    actor = unreal.EditorLevelLibrary.spawn_actor_from_class(
        unreal.StaticMeshActor,
        unreal.Vector(location[0] * 100.0, location[1] * 100.0, location[2] * 100.0),
    )
    actor.set_actor_label(name)
    actor.set_actor_scale3d(unreal.Vector(scale[0], scale[1], scale[2]))
    component = actor.get_component_by_class(unreal.StaticMeshComponent)
    if component is None:
        raise RuntimeError("StaticMeshActor has no StaticMeshComponent: " + name)
    component.set_static_mesh(load_mesh(geometry))
    if material is not None:
        component.set_material(0, material)
    return actor

def build_descriptor_parts():
    for part in scene.get("parts", []):
        material = material_for_descriptor(material_descriptor(part.get("material", "")))
        actor = spawn_actor(part.get("id", "part"), part.get("position", [0, 0, 0]), part.get("scale", [0.1, 0.1, 0.1]), part.get("geometry", "cube"), material)
        rotation = part.get("rotation", [])
        if len(rotation) >= 3:
            actor.set_actor_rotation(unreal.Rotator(math.degrees(rotation[1]), math.degrees(rotation[2]), math.degrees(rotation[0])), False)

def build_lineup(prefix, count):
    for index in range(count):
        x = -1.65 + index * 0.33
        y = math.sin(index * 0.8) * 0.18
        geometry = "sphere" if index % 3 != 0 else "cube"
        spawn_actor(prefix + str(index), [x, y, 0], [0.22, 0.22, 0.22], geometry, fallback_material(prefix + str(index), index))

def build_shadow():
    spawn_actor("receiver", [0, -0.55, 0.18], [2.1, 0.08, 0.8], "cube", fallback_material("shadow-receiver", 3))
    spawn_actor("caster-a", [-0.32, 0.05, 0], [0.38, 0.78, 0.38], "cube", fallback_material("shadow-caster-a", 4))
    spawn_actor("caster-b", [0.48, 0.02, 0], [0.48, 0.48, 0.48], "sphere", fallback_material("shadow-caster-b", 5))

if len(scene.get("parts", [])) > 0:
    build_descriptor_parts()
elif baseline_kind == "product-visual":
    build_descriptor_parts()
elif baseline_kind == "shadow-visual":
    build_shadow()
elif baseline_kind == "postprocess-suite":
    build_lineup("postprocess-sample-", 14)
else:
    minimum = scene.get("minimumEvidence", {})
    build_lineup(baseline_kind + "-sample-", int(minimum.get("materialCount", minimum.get("featureCount", 6))))

camera = unreal.EditorLevelLibrary.spawn_actor_from_class(unreal.CameraActor, unreal.Vector(0, 0, -600))
camera.set_actor_rotation(unreal.Rotator(0, 0, 0), False)
directional = unreal.EditorLevelLibrary.spawn_actor_from_class(unreal.DirectionalLight, unreal.Vector(-100, -200, -300))
directional.set_actor_rotation(unreal.Rotator(-42, -28, 0), False)
for light_index, light_color in enumerate([[1.0, 0.82, 0.62], [0.35, 0.62, 1.0], [0.34, 1.0, 0.58]]):
    light = unreal.EditorLevelLibrary.spawn_actor_from_class(unreal.PointLight, unreal.Vector((-160 + light_index * 160), 180, 160))
    component = light.get_component_by_class(unreal.PointLightComponent)
    if component is not None:
        component.set_editor_property("intensity", 550.0)
        component.set_editor_property("light_color", unreal.Color(int(light_color[0] * 255), int(light_color[1] * 255), int(light_color[2] * 255), 255))
try:
    unreal.EditorLevelLibrary.set_level_viewport_camera_info(camera.get_actor_location(), camera.get_actor_rotation())
except Exception as error:
    unreal.log_warning("Galileo3D V4 baseline could not set viewport camera automatically: " + str(error))

minimum = scene.get("minimumEvidence", {})
viewport = scene.get("viewport", {})
width = int(minimum.get("width", viewport.get("width", 720)))
height = int(minimum.get("height", viewport.get("height", 480)))
try:
    unreal.AutomationLibrary.take_high_res_screenshot(width, height, SCREENSHOT_PATH)
    unreal.log("Galileo3D V4 external baseline screenshot requested for " + baseline_kind + ": " + SCREENSHOT_PATH + " at " + str(width) + "x" + str(height))
except Exception as error:
    unreal.log_warning("Galileo3D V4 baseline automatic screenshot failed; capture the viewport manually to " + SCREENSHOT_PATH + ": " + str(error))
screenshot_captured = False
for _attempt in range(60):
    if os.path.exists(SCREENSHOT_PATH) and os.path.getsize(SCREENSHOT_PATH) > 0:
        screenshot_captured = True
        break
    time.sleep(1.0)
if not screenshot_captured:
    unreal.log_warning("Galileo3D V4 baseline screenshot was not present after waiting; runner evidence will remain non-capturing until a real PNG exists at " + SCREENSHOT_PATH)
evidence_path = SCREENSHOT_PATH + ".evidence.json"
evidence_directory = os.path.dirname(evidence_path)
if evidence_directory:
    os.makedirs(evidence_directory, exist_ok=True)
evidence_metrics = {
    "width": width,
    "height": height,
    "drawCalls": int(minimum.get("drawCalls", 1)),
    "materialCount": int(max(float(minimum.get("materialCount", 0)), len(scene.get("materials", [])))),
    "productParts": int(max(float(minimum.get("productParts", 0)), len(scene.get("parts", [])))),
    "turntableHotspots": int(minimum.get("turntableHotspots", 0)),
    "captureViews": int(minimum.get("captureViews", 0)),
    "batchTasks": int(minimum.get("batchTasks", 0)),
    "featureCount": int(minimum.get("featureCount", 0)),
    "shadowEvidencePixels": int(minimum.get("shadowEvidencePixels", 0)),
    "toneMappedPatches": int(minimum.get("toneMappedPatches", 0)),
    "implementedEffects": int(minimum.get("implementedEffects", 0)),
    "realSceneEffects": int(minimum.get("realSceneEffects", 0)),
}
runner_evidence = {
    "ok": screenshot_captured,
    "engine": "unreal",
    "baselineKind": baseline_kind,
    "sceneDescriptorId": scene.get("id", ""),
    "sceneDescriptorVersion": scene.get("schemaVersion", ""),
    "screenshotPath": SCREENSHOT_PATH,
    "renderedFrameCaptured": screenshot_captured,
    "cameraConfigured": True,
    "metrics": evidence_metrics,
    "claimBoundary": "Runner evidence proves the Unreal scaffold built the descriptor scene and waited for a rendered PNG. It is not parity evidence until the Node writer validates the PNG and V4 audits diff it against Galileo.",
}
with open(evidence_path, "w", encoding="utf-8") as evidence_file:
    json.dump(runner_evidence, evidence_file, indent=2)
    evidence_file.write("\n")
unreal.log("Galileo3D V4 external baseline runner evidence written: " + evidence_path)
unreal.log("Galileo3D V4 external baseline scene built for " + baseline_kind + ". Expected screenshot: " + SCREENSHOT_PATH)

File: release-artifacts/v4-external-evidence-handoff/fixtures/external-engine-baselines/v4/unity/V4ExternalVisualBaselineRunner.cs
using System;
using System.Collections;
using System.IO;
using System.Linq;
using UnityEngine;

public sealed class V4ExternalVisualBaselineRunner : MonoBehaviour
{
    // Assign one of the generated *-scene.json descriptors and set BaselineKind
    // to product-visual, pbr-visual, shadow-visual, hdr-render-target, or postprocess-suite.
    // This scaffold builds a deterministic approximation that must be captured by a
    // real Unity editor run before any Unity parity claim is allowed.
    public TextAsset SceneDescriptor;
    public string BaselineKind = "product-visual";
    public string ScreenshotPath = "tests/reports/v4-external-baseline/unity-baseline.png";

    public static void CaptureFromCommandLine()
    {
        var descriptorPath = CommandLineValue("--descriptor");
        var screenshotPath = CommandLineValue("--screenshot");
        var baselineKind = CommandLineValue("--baseline-kind");
        if (string.IsNullOrEmpty(descriptorPath))
        {
            throw new InvalidOperationException("Missing --descriptor <path> for V4 Unity baseline capture.");
        }
        if (string.IsNullOrEmpty(screenshotPath))
        {
            throw new InvalidOperationException("Missing --screenshot <path> for V4 Unity baseline capture.");
        }
        if (!File.Exists(descriptorPath))
        {
            throw new FileNotFoundException("V4 Unity baseline descriptor was not found.", descriptorPath);
        }
        var descriptor = JsonUtility.FromJson<BaselineSceneDescriptor>(File.ReadAllText(descriptorPath));
        var resolvedKind = string.IsNullOrEmpty(baselineKind) ? descriptor.baselineKind : baselineKind;
        CaptureDescriptor(descriptor, resolvedKind, screenshotPath);
        Debug.Log("Galileo3D V4 Unity baseline command-line capture completed for " + resolvedKind + ": " + screenshotPath);
    }

    private IEnumerator Start()
    {
        if (SceneDescriptor == null)
        {
            throw new InvalidOperationException("Assign a generated V4 external baseline scene descriptor.");
        }
        var descriptor = JsonUtility.FromJson<BaselineSceneDescriptor>(SceneDescriptor.text);
        var baselineKind = string.IsNullOrEmpty(descriptor.baselineKind) ? BaselineKind : descriptor.baselineKind;
        yield return new WaitForEndOfFrame();
        CaptureDescriptor(descriptor, baselineKind, ScreenshotPath);
    }

    private static void CaptureDescriptor(BaselineSceneDescriptor descriptor, string baselineKind, string screenshotPath)
    {
        var targetWidth = descriptor.TargetWidth();
        var targetHeight = descriptor.TargetHeight();
        if (targetWidth > 0 && targetHeight > 0)
        {
            Screen.SetResolution(targetWidth, targetHeight, false);
        }
        BuildScene(descriptor, baselineKind);
        var cameraObject = Camera.main != null ? Camera.main.gameObject : new GameObject("baseline-camera");
        var camera = cameraObject.GetComponent<Camera>() ?? cameraObject.AddComponent<Camera>();
        camera.clearFlags = CameraClearFlags.SolidColor;
        camera.backgroundColor = new Color(0.025f, 0.03f, 0.04f, 1);
        camera.orthographic = true;
        camera.orthographicSize = baselineKind == "pbr-visual" || baselineKind == "postprocess-suite" ? 2.2f : 1.35f;
        cameraObject.transform.position = new Vector3(0, 0, -6);
        cameraObject.transform.rotation = Quaternion.identity;
        var directory = Path.GetDirectoryName(screenshotPath);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }
        var screenshotBytes = CaptureCameraPng(camera, targetWidth, targetHeight, screenshotPath);
        if (screenshotBytes <= 0)
        {
            throw new InvalidOperationException("Synchronous Unity baseline capture wrote an empty PNG: " + screenshotPath);
        }
        WriteRunnerEvidence(descriptor, baselineKind, screenshotPath, targetWidth, targetHeight);
        Debug.Log("Galileo3D V4 external baseline screenshot captured: " + screenshotPath + " at " + targetWidth + "x" + targetHeight + " (" + screenshotBytes + " bytes)");
    }

    private static string CommandLineValue(string key)
    {
        var args = Environment.GetCommandLineArgs();
        for (var i = 0; i < args.Length - 1; i++)
        {
            if (args[i] == key) return args[i + 1];
        }
        return "";
    }

    private static void BuildScene(BaselineSceneDescriptor descriptor, string baselineKind)
    {
        AddLighting(baselineKind);
        if (descriptor.parts != null && descriptor.parts.Length > 0)
        {
            foreach (var part in descriptor.parts)
            {
                CreatePart(part, FindMaterial(descriptor.materials, part.material));
            }
            return;
        }
        switch (baselineKind)
        {
            case "pbr-visual":
                BuildPbrLineup();
                break;
            case "shadow-visual":
                BuildShadowScene();
                break;
            case "hdr-render-target":
                BuildHdrPatches();
                break;
            case "postprocess-suite":
                BuildPostprocessScene();
                break;
            default:
                BuildProductScene();
                break;
        }
    }

    private static void BuildProductScene()
    {
        var body = CreatePart("body", PrimitiveType.Cube, new Vector3(0, -0.04f, 0), new Vector3(1.44f, 0.76f, 0.36f), new Color(0.1f, 0.14f, 0.18f, 1), 0.78f, 0.24f);
        CreatePart("screen", PrimitiveType.Cube, new Vector3(0, 0.02f, 0.03f), new Vector3(1.04f, 0.48f, 0.06f), new Color(0.08f, 0.32f, 0.52f, 0.82f), 0.1f, 0.18f);
        CreatePart("left-dial", PrimitiveType.Sphere, new Vector3(-0.54f, -0.08f, 0), new Vector3(0.56f, 0.56f, 0.16f), Color.black, 0.4f, 0.35f);
        CreatePart("right-dial", PrimitiveType.Sphere, new Vector3(0.54f, -0.08f, 0), new Vector3(0.56f, 0.56f, 0.16f), Color.black, 0.4f, 0.35f);
        for (var i = 0; i < 14; i++)
        {
            var x = -0.84f + i * 0.13f;
            CreatePart("detail-" + i, PrimitiveType.Cube, new Vector3(x, -0.31f + (i % 3) * 0.2f, 0.08f), new Vector3(0.07f, 0.035f, 0.03f), new Color(0.9f, 0.48f, 0.18f, 1), 0.2f, 0.42f);
        }
    }

    private static void BuildPbrLineup()
    {
        for (var i = 0; i < 11; i++)
        {
            var x = -1.65f + i * 0.33f;
            var metallic = i / 10f;
            var roughness = 1f - i / 12f;
            CreatePart("pbr-sample-" + i, PrimitiveType.Sphere, new Vector3(x, 0, 0), new Vector3(0.22f, 0.22f, 0.22f), Color.HSVToRGB(i / 11f, 0.55f, 0.9f), metallic, roughness);
        }
    }

    private static void BuildShadowScene()
    {
        CreatePart("receiver", PrimitiveType.Cube, new Vector3(0, -0.55f, 0.18f), new Vector3(2.1f, 0.08f, 0.8f), new Color(0.72f, 0.74f, 0.68f, 1), 0.0f, 0.7f);
        CreatePart("caster-a", PrimitiveType.Cube, new Vector3(-0.32f, 0.05f, 0), new Vector3(0.38f, 0.78f, 0.38f), new Color(0.22f, 0.44f, 0.9f, 1), 0.15f, 0.38f);
        CreatePart("caster-b", PrimitiveType.Sphere, new Vector3(0.48f, 0.02f, 0), new Vector3(0.48f, 0.48f, 0.48f), new Color(0.9f, 0.42f, 0.18f, 1), 0.25f, 0.32f);
    }

    private static void BuildHdrPatches()
    {
        CreatePart("hdr-dark", PrimitiveType.Cube, new Vector3(-0.72f, 0, 0), new Vector3(0.45f, 0.62f, 0.12f), new Color(0.12f, 0.1f, 0.08f, 1), 0, 0.45f);
        CreatePart("hdr-mid", PrimitiveType.Cube, new Vector3(0, 0, 0), new Vector3(0.45f, 0.62f, 0.12f), new Color(0.85f, 0.62f, 0.25f, 1), 0.1f, 0.28f);
        CreatePart("hdr-hot", PrimitiveType.Cube, new Vector3(0.72f, 0, 0), new Vector3(0.45f, 0.62f, 0.12f), new Color(1f, 0.92f, 0.42f, 1), 0.0f, 0.12f);
    }

    private static void BuildPostprocessScene()
    {
        for (var i = 0; i < 14; i++)
        {
            var angle = i * Mathf.PI * 2f / 14f;
            var radius = 0.35f + (i % 3) * 0.2f;
            CreatePart("postprocess-sample-" + i, PrimitiveType.Sphere, new Vector3(Mathf.Cos(angle) * radius, Mathf.Sin(angle) * radius, 0), new Vector3(0.18f, 0.18f, 0.18f), Color.HSVToRGB(i / 14f, 0.7f, 1), i % 2 == 0 ? 0.35f : 0.05f, 0.25f + i * 0.025f);
        }
    }

    private static void AddLighting(string baselineKind)
    {
        var lightObject = new GameObject("v4-external-baseline-key-light");
        var light = lightObject.AddComponent<Light>();
        light.type = LightType.Directional;
        light.intensity = baselineKind == "hdr-render-target" ? 1.6f : 1.1f;
        lightObject.transform.rotation = Quaternion.Euler(38, -32, 0);
    }

    private static GameObject CreatePart(string name, PrimitiveType primitive, Vector3 position, Vector3 scale, Color color, float metallic, float roughness)
    {
        var obj = GameObject.CreatePrimitive(primitive);
        obj.name = name;
        obj.transform.position = position;
        obj.transform.localScale = scale;
        var material = new Material(Shader.Find("Standard"));
        material.color = color;
        material.SetFloat("_Metallic", metallic);
        material.SetFloat("_Glossiness", 1f - roughness);
        obj.GetComponent<Renderer>().material = material;
        return obj;
    }

    private static GameObject CreatePart(PartDescriptor part, MaterialDescriptor material)
    {
        var obj = GameObject.CreatePrimitive(PrimitiveFromString(part.geometry));
        obj.name = string.IsNullOrEmpty(part.id) ? "baseline-part" : part.id;
        obj.transform.position = ToVector3(part.position, Vector3.zero);
        obj.transform.localScale = ToVector3(part.scale, Vector3.one);
        if (part.rotation != null && part.rotation.Length >= 3)
        {
            obj.transform.rotation = Quaternion.Euler(part.rotation[0] * Mathf.Rad2Deg, part.rotation[1] * Mathf.Rad2Deg, part.rotation[2] * Mathf.Rad2Deg);
        }
        obj.GetComponent<Renderer>().material = CreateMaterial(material);
        return obj;
    }

    private static MaterialDescriptor FindMaterial(MaterialDescriptor[] materials, string id)
    {
        if (materials != null)
        {
            foreach (var material in materials)
            {
                if (material != null && material.id == id) return material;
            }
        }
        return new MaterialDescriptor { id = id, kind = "pbr", color = new[] { 0.78f, 0.78f, 0.78f, 1f }, metallic = 0.0f, roughness = 0.5f };
    }

    private static Material CreateMaterial(MaterialDescriptor descriptor)
    {
        var material = new Material(Shader.Find(descriptor.kind == "unlit" ? "Unlit/Color" : "Standard"));
        var color = ToColor(descriptor.color, Color.white);
        material.color = color;
        if (descriptor.kind != "unlit")
        {
            material.SetFloat("_Metallic", descriptor.metallic);
            material.SetFloat("_Glossiness", 1f - descriptor.roughness);
        }
        return material;
    }

    private static PrimitiveType PrimitiveFromString(string geometry)
    {
        switch (geometry)
        {
            case "sphere":
                return PrimitiveType.Sphere;
            case "cylinder":
                return PrimitiveType.Cylinder;
            default:
                return PrimitiveType.Cube;
        }
    }

    private static Vector3 ToVector3(float[] values, Vector3 fallback)
    {
        if (values == null || values.Length < 3) return fallback;
        return new Vector3(values[0], values[1], values[2]);
    }

    private static Color ToColor(float[] values, Color fallback)
    {
        if (values == null || values.Length < 3) return fallback;
        return new Color(values[0], values[1], values[2], values.Length >= 4 ? values[3] : 1f);
    }

    private static int CaptureCameraPng(Camera camera, int width, int height, string screenshotPath)
    {
        var captureWidth = Mathf.Max(1, width);
        var captureHeight = Mathf.Max(1, height);
        var previousTarget = camera.targetTexture;
        var previousActive = RenderTexture.active;
        var renderTexture = new RenderTexture(captureWidth, captureHeight, 24, RenderTextureFormat.ARGB32);
        var texture = new Texture2D(captureWidth, captureHeight, TextureFormat.RGBA32, false);
        try
        {
            camera.targetTexture = renderTexture;
            RenderTexture.active = renderTexture;
            camera.Render();
            texture.ReadPixels(new Rect(0, 0, captureWidth, captureHeight), 0, 0);
            texture.Apply(false, false);
            var bytes = texture.EncodeToPNG();
            File.WriteAllBytes(screenshotPath, bytes);
            return bytes.Length;
        }
        finally
        {
            camera.targetTexture = previousTarget;
            RenderTexture.active = previousActive;
            if (Application.isPlaying)
            {
                Destroy(texture);
                renderTexture.Release();
                Destroy(renderTexture);
            }
            else
            {
                DestroyImmediate(texture);
                renderTexture.Release();
                DestroyImmediate(renderTexture);
            }
        }
    }

    private static void WriteRunnerEvidence(BaselineSceneDescriptor descriptor, string baselineKind, string screenshotPath, int width, int height)
    {
        var metrics = RunnerEvidenceMetrics.FromDescriptor(descriptor, width, height);
        var report = new RunnerEvidenceReport
        {
            ok = true,
            engine = "unity",
            baselineKind = baselineKind,
            sceneDescriptorId = descriptor.id,
            sceneDescriptorVersion = descriptor.schemaVersion,
            screenshotPath = screenshotPath,
            renderedFrameCaptured = true,
            cameraConfigured = true,
            metrics = metrics,
            claimBoundary = "Runner evidence proves the Unity scaffold built the descriptor scene and synchronously wrote a rendered camera PNG. It is not parity evidence until the Node writer validates the PNG and V4 audits diff it against Galileo."
        };
        var evidencePath = screenshotPath + ".evidence.json";
        var directory = Path.GetDirectoryName(evidencePath);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }
        File.WriteAllText(evidencePath, JsonUtility.ToJson(report, true));
    }

    [Serializable]
    private sealed class BaselineSceneDescriptor
    {
        public string id = "";
        public string schemaVersion = "";
        public string baselineKind = "";
        public ViewportDescriptor viewport = new ViewportDescriptor();
        public MinimumEvidenceDescriptor minimumEvidence = new MinimumEvidenceDescriptor();
        public MaterialDescriptor[] materials = new MaterialDescriptor[0];
        public PartDescriptor[] parts = new PartDescriptor[0];

        public int TargetWidth()
        {
            if (minimumEvidence != null && minimumEvidence.width > 0) return minimumEvidence.width;
            return viewport != null ? viewport.width : 0;
        }

        public int TargetHeight()
        {
            if (minimumEvidence != null && minimumEvidence.height > 0) return minimumEvidence.height;
            return viewport != null ? viewport.height : 0;
        }
    }

    [Serializable]
    private sealed class ViewportDescriptor
    {
        public int width = 0;
        public int height = 0;
    }

    [Serializable]
    private sealed class MinimumEvidenceDescriptor
    {
        public int width = 0;
        public int height = 0;
        public int drawCalls = 0;
        public int materialCount = 0;
        public int productParts = 0;
        public int turntableHotspots = 0;
        public int captureViews = 0;
        public int batchTasks = 0;
        public int featureCount = 0;
        public int shadowEvidencePixels = 0;
        public int toneMappedPatches = 0;
        public int implementedEffects = 0;
        public int realSceneEffects = 0;
    }

    [Serializable]
    private sealed class RunnerEvidenceReport
    {
        public bool ok = true;
        public string engine = "unity";
        public string baselineKind = "";
        public string sceneDescriptorId = "";
        public string sceneDescriptorVersion = "";
        public string screenshotPath = "";
        public bool renderedFrameCaptured = true;
        public bool cameraConfigured = true;
        public RunnerEvidenceMetrics metrics = new RunnerEvidenceMetrics();
        public string claimBoundary = "";
    }

    [Serializable]
    private sealed class RunnerEvidenceMetrics
    {
        public int width = 0;
        public int height = 0;
        public int drawCalls = 0;
        public int materialCount = 0;
        public int productParts = 0;
        public int turntableHotspots = 0;
        public int captureViews = 0;
        public int batchTasks = 0;
        public int featureCount = 0;
        public int shadowEvidencePixels = 0;
        public int toneMappedPatches = 0;
        public int implementedEffects = 0;
        public int realSceneEffects = 0;

        public static RunnerEvidenceMetrics FromDescriptor(BaselineSceneDescriptor descriptor, int width, int height)
        {
            var minimum = descriptor.minimumEvidence ?? new MinimumEvidenceDescriptor();
            return new RunnerEvidenceMetrics
            {
                width = width,
                height = height,
                drawCalls = Math.Max(1, minimum.drawCalls),
                materialCount = Math.Max(minimum.materialCount, descriptor.materials == null ? 0 : descriptor.materials.Length),
                productParts = Math.Max(minimum.productParts, descriptor.parts == null ? 0 : descriptor.parts.Length),
                turntableHotspots = minimum.turntableHotspots,
                captureViews = minimum.captureViews,
                batchTasks = minimum.batchTasks,
                featureCount = minimum.featureCount,
                shadowEvidencePixels = minimum.shadowEvidencePixels,
                toneMappedPatches = minimum.toneMappedPatches,
                implementedEffects = minimum.implementedEffects,
                realSceneEffects = minimum.realSceneEffects
            };
        }
    }

    [Serializable]
    private sealed class MaterialDescriptor
    {
        public string id = "";
        public string kind = "pbr";
        public float[] color = new[] { 1f, 1f, 1f, 1f };
        public float metallic = 0f;
        public float roughness = 0.5f;
    }

    [Serializable]
    private sealed class PartDescriptor
    {
        public string id = "";
        public string geometry = "cube";
        public string material = "";
        public float[] position = new[] { 0f, 0f, 0f };
        public float[] scale = new[] { 1f, 1f, 1f };
        public float[] rotation = new float[0];
    }
}
